File: dns_lookup.go
// dns_lookup.go — Go версия

package main

import (
	"bufio"
	"encoding/csv"
	"encoding/json"
	"fmt"
	"net"
	"os"
	"strings"
	"sync"
	"time"
)

type DNSResult struct {
	Domain    string   `json:"domain"`
	Timestamp string   `json:"timestamp"`
	AAAA      []string `json:"AAAA"`
	PTR       []string `json:"PTR"`
}

type DNSLookup struct {
	timeout time.Duration
	threads int
	results []DNSResult
	mu      sync.Mutex
}

func NewDNSLookup(timeout int, threads int) *DNSLookup {
	return &DNSLookup{
		timeout: time.Duration(timeout) * time.Second,
		threads: threads,
		results: []DNSResult{},
	}
}

func (d *DNSLookup) lookupAAAA(domain string) []string {
	ips, err := net.LookupIP(domain)
	if err != nil {
		return []string{}
	}
	var result []string
	for _, ip := range ips {
		if ip.To4() == nil && ip.To16() != nil {
			result = append(result, ip.String())
		}
	}
	return result
}

func (d *DNSLookup) lookupPTR(ip string) []string {
	names, err := net.LookupAddr(ip)
	if err != nil {
		return []string{}
	}
	return names
}

func (d *DNSLookup) lookupDomain(domain string) DNSResult {
	domain = strings.TrimSpace(domain)
	if domain == "" {
		return DNSResult{}
	}

	// Проверяем, является ли строка IPv6-адресом
	if ip := net.ParseIP(domain); ip != nil && ip.To4() == nil && ip.To16() != nil {
		ptr := d.lookupPTR(domain)
		return DNSResult{
			Domain:    domain,
			Timestamp: time.Now().Format(time.RFC3339),
			AAAA:      []string{},
			PTR:       ptr,
		}
	}

	aaaa := d.lookupAAAA(domain)
	var ptr []string
	if len(aaaa) > 0 {
		for _, ip := range aaaa {
			names := d.lookupPTR(ip)
			ptr = append(ptr, names...)
		}
	}
	if len(ptr) == 0 {
		ptr = []string{"не найден"}
	}

	return DNSResult{
		Domain:    domain,
		Timestamp: time.Now().Format(time.RFC3339),
		AAAA:      aaaa,
		PTR:       ptr,
	}
}

func (d *DNSLookup) lookupBatch(domains []string) {
	fmt.Printf("🔍 Выполняем lookup для %d доменов...\n", len(domains))
	start := time.Now()

	var wg sync.WaitGroup
	sem := make(chan struct{}, d.threads)

	for _, domain := range domains {
		if domain == "" {
			continue
		}
		wg.Add(1)
		go func(dmn string) {
			defer wg.Done()
			sem <- struct{}{}
			defer func() { <-sem }()

			result := d.lookupDomain(dmn)
			d.mu.Lock()
			d.results = append(d.results, result)
			d.mu.Unlock()
			fmt.Printf("✅ %s — выполнено\n", dmn)
		}(domain)
	}

	wg.Wait()
	fmt.Printf("⏱️ Время выполнения: %.2f сек.\n", time.Since(start).Seconds())
}

func (d *DNSLookup) saveJSON(filename string) error {
	data, err := json.MarshalIndent(d.results, "", "  ")
	if err != nil {
		return err
	}
	return os.WriteFile(filename, data, 0644)
}

func (d *DNSLookup) saveCSV(filename string) error {
	file, err := os.Create(filename)
	if err != nil {
		return err
	}
	defer file.Close()

	writer := csv.NewWriter(file)
	defer writer.Flush()

	writer.Write([]string{"Domain", "Timestamp", "AAAA", "PTR"})
	for _, r := range d.results {
		writer.Write([]string{
			r.Domain,
			r.Timestamp,
			strings.Join(r.AAAA, "; "),
			strings.Join(r.PTR, "; "),
		})
	}
	return nil
}

func (d *DNSLookup) printTable() {
	if len(d.results) == 0 {
		fmt.Println("Нет результатов.")
		return
	}

	fmt.Println("\n" + strings.Repeat("=", 90))
	fmt.Printf("%-25s %-35s %-25s\n", "Домен/IP", "AAAA", "PTR")
	fmt.Println(strings.Repeat("-", 90))
	for _, r := range d.results {
		aaaa := strings.Join(r.AAAA, "; ")
		if len(aaaa) > 35 {
			aaaa = aaaa[:35] + "..."
		}
		ptr := strings.Join(r.PTR, "; ")
		if len(ptr) > 25 {
			ptr = ptr[:25] + "..."
		}
		fmt.Printf("%-25s %-35s %-25s\n", r.Domain, aaaa, ptr)
	}
	fmt.Println(strings.Repeat("=", 90))
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("Usage: go run dns_lookup.go <domains.txt>")
		fmt.Println("   или: go run dns_lookup.go domain1.com domain2.com")
		os.Exit(1)
	}

	var domains []string

	if len(os.Args) == 2 && !strings.Contains(os.Args[1], ".") {
		file, err := os.Open(os.Args[1])
		if err == nil {
			defer file.Close()
			scanner := bufio.NewScanner(file)
			for scanner.Scan() {
				line := strings.TrimSpace(scanner.Text())
				if line != "" {
					domains = append(domains, line)
				}
			}
		}
	} else {
		for _, arg := range os.Args[1:] {
			domains = append(domains, arg)
		}
	}

	if len(domains) == 0 {
		fmt.Println("❌ Нет доменов для проверки.")
		os.Exit(1)
	}

	fmt.Println("🌐 DNS Lookup IPv6 Pro (Go)")
	fmt.Printf("📂 Загружено %d доменов.\n", len(domains))

	lookup := NewDNSLookup(5, 20)
	lookup.lookupBatch(domains)

	if len(lookup.results) > 0 {
		lookup.printTable()
		lookup.saveJSON("dns_results.json")
		lookup.saveCSV("dns_results.csv")
		fmt.Println("💾 Сохранено: dns_results.json, dns_results.csv")
	} else {
		fmt.Println("❌ Не удалось выполнить lookup.")
	}
}

File: dns_lookup.js
// dns_lookup.js — JavaScript версия

const dns = require('dns');
const fs = require('fs');
const path = require('path');
const { promisify } = require('util');

const resolve4 = promisify(dns.resolve4);
const resolve6 = promisify(dns.resolve6);
const resolvePtr = promisify(dns.resolvePtr);
const lookup = promisify(dns.lookup);

class DNSLookupIPv6 {
    constructor(options = {}) {
        this.timeout = options.timeout || 5000;
        this.threads = options.threads || 20;
        this.results = [];
    }

    isIPv6(addr) {
        return /^([0-9a-fA-F]{1,4}:){7}[0-9a-fA-F]{1,4}$|^::1$|^fe80::/.test(addr);
    }

    async lookupAAAA(domain) {
        try {
            const ips = await resolve6(domain);
            return ips;
        } catch (err) {
            return [];
        }
    }

    async lookupPTR(ip) {
        try {
            const names = await resolvePtr(ip);
            return names;
        } catch (err) {
            return [];
        }
    }

    async lookupDomain(domain) {
        domain = domain.trim();
        if (!domain) return null;

        // Проверяем, является ли строка IPv6-адресом
        if (this.isIPv6(domain)) {
            const ptr = await this.lookupPTR(domain);
            return {
                domain: domain,
                timestamp: new Date().toISOString(),
                AAAA: [],
                PTR: ptr.length ? ptr : ['не найден']
            };
        }

        const aaaa = await this.lookupAAAA(domain);
        let ptr = [];
        if (aaaa.length > 0) {
            for (const ip of aaaa) {
                const names = await this.lookupPTR(ip);
                ptr = ptr.concat(names);
            }
        }
        if (ptr.length === 0) {
            ptr = ['не найден'];
        }

        return {
            domain: domain,
            timestamp: new Date().toISOString(),
            AAAA: aaaa,
            PTR: ptr
        };
    }

    async lookupBatch(domains) {
        console.log(`🔍 Выполняем lookup для ${domains.length} доменов...`);
        const start = Date.now();

        const chunks = [];
        for (let i = 0; i < domains.length; i += this.threads) {
            chunks.push(domains.slice(i, i + this.threads));
        }

        for (const chunk of chunks) {
            const promises = chunk.map(async domain => {
                const result = await this.lookupDomain(domain);
                if (result) {
                    this.results.push(result);
                    console.log(`✅ ${domain} — выполнено`);
                }
            });
            await Promise.all(promises);
        }

        console.log(`⏱️ Время выполнения: ${(Date.now() - start) / 1000} сек.`);
    }

    saveJSON(filename) {
        fs.writeFileSync(filename, JSON.stringify(this.results, null, 2));
        console.log(`💾 Сохранено JSON: ${filename}`);
    }

    saveCSV(filename) {
        if (this.results.length === 0) {
            console.log('⚠️ Нет данных для сохранения.');
            return;
        }
        const headers = ['Domain', 'Timestamp', 'AAAA', 'PTR'];
        const rows = this.results.map(r => [
            r.domain,
            r.timestamp,
            r.AAAA.join('; ') || '—',
            r.PTR.join('; ') || 'не найден'
        ]);
        const csv = [headers.join(','), ...rows.map(row => row.join(','))].join('\n');
        fs.writeFileSync(filename, csv);
        console.log(`💾 Сохранено CSV: ${filename}`);
    }

    printTable() {
        if (this.results.length === 0) {
            console.log('Нет результатов.');
            return;
        }
        console.log('\n' + '='.repeat(90));
        console.log(`${'Домен/IP'.padEnd(25)} ${'AAAA'.padEnd(35)} ${'PTR'.padEnd(25)}`);
        console.log('-'.repeat(90));
        for (const r of this.results) {
            let aaaa = r.AAAA.join('; ') || '—';
            let ptr = r.PTR.join('; ') || 'не найден';
            if (aaaa.length > 35) aaaa = aaaa.slice(0, 35) + '...';
            if (ptr.length > 25) ptr = ptr.slice(0, 25) + '...';
            console.log(`${r.domain.padEnd(25)} ${aaaa.padEnd(35)} ${ptr.padEnd(25)}`);
        }
        console.log('='.repeat(90));
    }
}

async function main() {
    const args = process.argv.slice(2);
    if (args.length === 0) {
        console.log('Usage: node dns_lookup.js <domains.txt>');
        console.log('   или: node dns_lookup.js domain1.com domain2.com');
        process.exit(1);
    }

    let domains = [];

    if (args.length === 1 && fs.existsSync(args[0])) {
        const content = fs.readFileSync(args[0], 'utf-8');
        domains = content.split('\n').map(l => l.trim()).filter(l => l);
    } else {
        domains = args;
    }

    if (domains.length === 0) {
        console.log('❌ Нет доменов для проверки.');
        process.exit(1);
    }

    console.log('🌐 DNS Lookup IPv6 Pro (JavaScript)');
    console.log(`📂 Загружено ${domains.length} доменов.`);

    const lookup = new DNSLookupIPv6({ timeout: 5000, threads: 20 });
    await lookup.lookupBatch(domains);

    if (lookup.results.length > 0) {
        lookup.printTable();
        lookup.saveJSON('dns_results.json');
        lookup.saveCSV('dns_results.csv');
    } else {
        console.log('❌ Не удалось выполнить lookup.');
    }
}

main().catch(console.error);
